### PR TITLE
nom 7.0 and code cleanup, fixes #21

### LINKED ISSUE #21: Utf8Error when trying to parse 2902.dat
I was looking for robust and efficient ldraw parsers in Rust and found this crate. From my initial experience, the library has been very well documented and easy to use. I noticed in #2 that the code likely hasn't been run against the entire parts library yet. I've already found some panics.

I'm getting the error `Utf8Error { valid_up_to: 10, error_len: Some(1) }` at this section of code. Rust doesn't seem to complain about the formatting when I trying and convert the file to utf8 with `std::fs::read_to_string`. This happens on the recently downloaded ldraw as well as bricklink studio parts libraries. I can upload the dat file if needed.
https://github.com/djeedai/weldr/blob/aae1bff12b85b7645be3983a68285149383b161e/lib/src/lib.rs#L414-L420

### PR DESCRIPTION
**Pull request by ScanMountGoat**
This PR reworks the parsing code to use functions instead of macros to support using the latest version of nom. Parsers that rely on the input being valid utf8 now return an error on invalid input to avoid panics. This fixes the panic in #21. I've also run `cargo fmt` and `cargo clippy --fix` on the code. I haven't fixed any dead code warnings to avoid deleting any WIP code.

### CODE REVIEW

**Review by djeedai — APPROVED** [review]
Looks all good, thanks a lot for that!

**Comment by djeedai** [comment]
I had to make some fixes on CI because 1.44 is a bit too old. Could you please merge latest `main` into your branch @ScanMountGoat? That should fix CI.

**Comment by ScanMountGoat** [comment]
> I had to make some fixes on CI because 1.44 is a bit too old. Could you please merge latest `main` into your branch @ScanMountGoat? That should fix CI.

I've updated my fork with the new commit.

**Comment by djeedai** [comment]
Another thing @ScanMountGoat is that I realize I originally copied `nom` and licensed under MIT alone, but I'd like to relicense to "MIT or Apache 2.0" for consistency with Rust and most of its projects (and my other crates). Technically I'm so far the only contributor until this PR merges, but I don't want this sudden change to come off as aggressive or a blocker for you and your work. If that's fine with you, I'll push the relicense first and we can merge this PR after. Then I'll maybe do a new release with the new license and those fixes. If there's any objection please let me know.

**Comment by djeedai** [comment]
See #24 for the actual relicensing. I'll wait a bit for an answer. Thanks!

**Comment by ScanMountGoat** [comment]
> Another thing @ScanMountGoat is that I realize I originally copied `nom` and licensed under MIT alone, but I'd like to relicense to "MIT or Apache 2.0" for consistency with Rust and most of its projects (and my other crates). Technically I'm so far the only contributor until this PR merges, but I don't want this sudden change to come off as aggressive or a blocker for you and your work. If that's fine with you, I'll push the relicense first and we can merge this PR after. Then I'll maybe do a new release with the new license and those fixes. If there's any objection please let me know.

I've read through the changes. I'm fine with my contributions falling under the new dual license. I agree that merging the license PR first is the right choice to avoid any ambiguities. Thanks for the heads up.

**Review by djeedai — APPROVED** [review]

**Pull request merged**
